src/ApiCore/ResourceTemplate/RelativeResourceTemplate.php

```diff
@@ -0,0 +1,296 @@
+        }
+
+        // Collapse the bindings from lists into strings
+        $collapsedBindings = [];
+        foreach ($bindings as $key => $boundPieces) {
+            $collapsedBindings[$key] = implode('/', $boundPieces);
+        }
+        return $collapsedBindings;
+    }
+
+    private function matchException($path)
+    {
+        return new ValidationException("Could not match path '$path' to template '$this'");
+    }
+
+    /**
+     * @param Segment[] $segments
+     * @return array[] A list of [string, Segment] tuples
+     */
+    private static function buildKeySegmentTuples($segments)
+    {
+        $keySegmentTuples = [];
+        $positionalArgumentCounter = 0;
+        foreach ($segments as $segment) {
+            switch ($segment->getSegmentType()) {
+                case Segment::WILDCARD_SEGMENT:
+                case Segment::DOUBLE_WILDCARD_SEGMENT:
+                    $positionalKey = "\$$positionalArgumentCounter";
+                    $positionalArgumentCounter++;
+                    $keySegmentTuples[] = [$positionalKey, $segment];
+                    break;
+                default:
+                    $keySegmentTuples[] = [$segment->getKey(), $segment];
+            }
+        }
+        return $keySegmentTuples;
+    }
+
+    /**
+     * @param array[] A list of [string, Segment] tuples
+     * @return array[] A list of [string, Segment] tuples
+     */
+    private static function flattenKeySegmentTuples($keySegmentTuples)
+    {
+        $flattenedKeySegmentTuples = [];
+        foreach ($keySegmentTuples as list($key, $segment)) {
+            /** @var Segment $segment */
+            switch ($segment->getSegmentType()) {
+                case Segment::VARIABLE_SEGMENT:
+                    // For segment variables, replace the segment with the segments of its children
+                    $template = $segment->getTemplate();
+                    $nestedKeySegmentTuples = self::buildKeySegmentTuples($template->segments);
+                    foreach ($nestedKeySegmentTuples as list($nestedKey, $nestedSegment)) {
+                        /** @var Segment $nestedSegment */
+                        // Nested variables are not allowed
+                        assert($nestedSegment->getSegmentType() !== Segment::VARIABLE_SEGMENT);
+                        // Insert the nested segment with key set to the outer key of the
+                        // parent variable segment
+                        $flattenedKeySegmentTuples[] = [$key, $nestedSegment];
+                    }
+                    break;
+                default:
+                    // For all other segments, don't change the key or segment
+                    $flattenedKeySegmentTuples[] = [$key, $segment];
+            }
+        }
+        return $flattenedKeySegmentTuples;
+    }
+
+    /**
+     * @param Segment[] $segments
+     * @return int
+     */
+    private static function countDoubleWildcards($segments)
```

Comment:
ditto

Reply:
Fixed in https://github.com/googleapis/gax-php/pull/209